build/commands/lib/actionGuard.js

```diff
@@ -18,32 +20,53 @@ function getGuardCallStack() {
 }
 
 // This class is used to ensure that a given action is successfully completed,
-// otherwise a cleanup might be performed.
+// otherwise a rerun might be required.
 class ActionGuard {
+  // Path to the guard file.
+  #guardFilePath
+  // Closure to perform the cleanup (optional).
+  #cleanupClosure
+  // Flag to indicate if the action is currently running.
+  #isRunning
+
   constructor(guardFilePath, cleanupClosure) {
-    this.guardFilePath = guardFilePath
-    this.cleanupClosure = cleanupClosure
+    this.#guardFilePath = guardFilePath
+    this.#cleanupClosure = cleanupClosure
+    this.#isRunning = false
   }
 
   // Check if the last action was interrupted.
-  isDirty() {
-    return fs.existsSync(this.guardFilePath)
-  }
-
-  // If the last action was interrupted, perform the cleanup and then perform
-  // the requested action with a guard.
-  ensureClean(actionClosure) {
-    if (this.isDirty()) {
-      this.cleanupClosure()
-    }
-    this.run(actionClosure)
+  wasInterrupted() {
+    assert(
+      !this.#isRunning,
+      'Cannot check if the action was interrupted while it is running.'
+    )
+    return fs.existsSync(this.#guardFilePath)
   }
 
   // Perform the requested action with a guard.
   run(actionClosure) {
-    fs.writeFileSync(this.guardFilePath, getGuardCallStack())
-    actionClosure()
-    fs.unlinkSync(this.guardFilePath)
+    assert(
+      !this.#isRunning,
+      'Cannot run the action while it is already running.'
+    )
+    const wasInterrupted = this.wasInterrupted()
+    if (wasInterrupted && this.#cleanupClosure) {
+      this.#cleanupClosure()
```

Comment:
if we pass `wasInterrupted` to `actionClosure`, we could perform cleanup inside the action.
Do we really need `cleanupClosure`?

Reply:
> if we pass `wasInterrupted` to `actionClosure`, we could perform cleanup inside the action. Do we really need `cleanupClosure`?

the guard file is checked after a cleanup by the guard. It should stay, otherwise we may end up in a broken state.